src/main/java/com/playmonumenta/networkrelay/BungeeNetworkMessageListener.java

```diff
@@ -53,4 +77,89 @@ public void networkRelayMessageEvent(NetworkRelayMessageEventBungee event) {
 
 		ProxyServer.getInstance().getPluginManager().dispatchCommand(ProxyServer.getInstance().getConsole(), data.get("command").getAsString());
 	}
+
+	@EventHandler(priority = EventPriority.NORMAL)
+	public void destOnline(DestOnlineEventBungee event) {
+		if (!mAutoRegisterBungeeServers) {
+			return;
+		}
+
+		String name = event.getDest();
+
+		JsonObject data = NetworkRelayAPI.getHeartbeatPluginData(name, NetworkRelayAPI.NETWORK_RELAY_HEARTBEAT_IDENTIFIER);
+
+		if (data != null && data.has("is-bungee")) {
+			JsonElement element = data.get("is-bungee");
+			if (element.isJsonPrimitive() && element.getAsJsonPrimitive().isBoolean()) {
+				boolean isBungee = element.getAsBoolean();
+				if (isBungee) {
+					// Don't want to add bungee proxies themselves as servers
+					return;
+				}
+			}
+		}
+
+		String serverAddress = null;
+		if (data != null && data.has("server-address")) {
+			JsonElement element = data.get("server-address");
+			if (element.isJsonPrimitive() && element.getAsJsonPrimitive().isString()) {
+				String addr = element.getAsString();
+				if (addr != null && !addr.isEmpty()) {
+					serverAddress = addr;
+				}
+			}
+		}
+		if (serverAddress == null) {
+			mLogger.warning("auto-register-bungee-servers=true but shard '" + name + "' is misconfigured and didn't send a server address");
+			return;
+		}
+
+		if (getServers().containsKey(name)) {
+			mLogger.warning("Tried to add server '" + name + "' due to heartbeat auto-registration but it already exists");
```

Comment:
What happens if a server crashes and restarts before it's detected as offline?

Reply:
Adding to this, same question if `mAutoUnregisterInactiveBungeeServers` is false, so it's never unregistered in the first place.

Reply:
Nevermind on that last question, I see the code that reads that config doesn't allow it to happen.

Reply:
Good question - fortunately this doesn't seem to cause any trouble. I don't think we'll actually use the unregistration at all. Multiple registrations are fine, it just does nothing (except print that log message)